### PR TITLE
optimizer: supports callsite annotations of inlining, fixes #18773

### LINKED ISSUE #18773: callsite inlining
Couldn't  find an issue  (somewhat related https://github.com/JuliaLang/julia/issues/15942) 

It would be great to be able to inline functions at callsite

``` julia
function f(x)
     y = x + 2
     @inline z = g(y)
     return z + y
end
```

This would be nice since it prevents having to rewrite the function `g(x)` to be a  'private' function that is always inlined with `@inline`   and then having to  rewriting a corresponding public function `g_public(x)` that  just calls private `g(x)` function that is always inlined.

### PR DESCRIPTION
**Pull request by aviatesk**
Enable `@inline`/`@noinline` annotations on function callsites.
From #40754.

Now `@inline` and `@noinline` can be applied to a code block and then
the compiler will try to (or try not to) inline calls within the block:
```julia
@inline f(...) # The compiler will try to inline `f`

@inline f(...) + g(...) # The compiler will try to inline `f`, `g` and `+`

@inline f(args...) = ... # Of course annotations on a definition is still allowed
```

Here are couple of notes on how those callsite annotations will work:
- callsite annotation always has the precedence over the annotation
  applied to the definition of the called method, whichever we use
  `@inline`/`@noinline`:
  ```julia
  @inline function explicit_inline(args...)
      # body
  end

  let
      @noinline explicit_inline(args...) # this call will not be inlined
  end
  ```
- when callsite annotations are nested, the innermost annotations has
  the precedence
  ```julia
  @noinline let a0, b0 = ...
      a = @inline f(a0)  # the compiler will try to inline this call
      b = notinlined(b0) # the compiler will NOT try to inline this call
      return a, b
  end
  ```
They're both tested and included in documentations.

Co-authored-by: Joseph Tan <jdtan638@gmail.com>

---

- closes #18773
- closes #40754

### CODE REVIEW

File: base/compiler/optimize.jl
Comment (aviatesk) [inline base/compiler/optimize.jl:47]: I tried the approach to use uncached source when the callsite is forced to be inlined (a1b1435f84f5891945dc233d73e606c69155098c), but it turns out that then inlining will depend o local state of inference and so the first entry and the succeeding ones may generate different code, e.g. from a test case I added:
```julia
function multilimited(a)
    if @noinline(isType(a))
        return @inline(multilimited(a.parameters[1]))
    else
        return rand(Bool) ? rand(a) : @inline(multilimited(a))
    end
end

# output might be different if we depend on inference local state
code_typed(m.multilimited, (Any,))
code_typed(m.multilimited, (Any,))
```

We might be able to keep source so that it doesn't depend on the local state, but it could be tricky.
Rather I'd like to always just re-infer when a source is unavailable, at least in this first iteration.

/cc @vtjnash

**Comment by tisztamo** [comment]
Does callsite `@inline` force inlining completely, ignoring the cost model? (also close #40292?)

**Comment by tisztamo** [comment]
Does callsite `@inline` force inlining completely, ignoring the cost model? (Also close #40292?) (re-commenting after accidental delete)

**Comment by aviatesk** [comment]
> Does callsite @inline force inlining completely, ignoring the cost model? 

It will ignore the cost model, but not "force inlining completely", since there are some cases when a method body itself is unavailable (think when there are recursive calls but annotated as `@inline`).

I would appreciate if you could test this PR for your target, though.

**Comment by tisztamo** [comment]
Just tested and it works, so this also closes #40292. Thank you very much!

File: base/compiler/optimize.jl
Comment (vtjnash) [reply base/compiler/optimize.jl:47]: I think this needs to happen during inference. It seems quite wrong to make a new NativeInterpreter here, and can probably have many other problems. We used to do some amount of recursion here, and we have gotten rid of it.

File: base/compiler/optimize.jl
Comment (vtjnash) [inline base/compiler/optimize.jl:424]: I wonder if it would be better process all of these flags better during `jl_code_info_set_ir`, since we already pre-process a number of other flags there, as part of converting from unattached lowered form to associated source code form.

Comment (aviatesk) [reply base/compiler/optimize.jl:424]: Can we just pack it into `CodeInfo.ssaflags` ?

Comment (vtjnash) [reply base/compiler/optimize.jl:423]: Yes. We currently don't.

Comment (aviatesk) [reply base/compiler/optimize.jl:387]: what do you think on ff0b8a9cdb90568d34a1a42180bb6a5877e69b30 ?

File: src/method.c
Comment (vtjnash) [inline src/method.c:321]:
```suggestion
            bd[j] = jl_nothing;
        }
```
usually after processing metadata, we remove it

File: src/method.c
Comment (vtjnash) [inline src/method.c:358]:
```suggestion
            jl_array_uint8_set(li->ssaflags, j, flag);
```

File: src/method.c
Comment (aviatesk) [reply src/method.c:325]: I think it's still useful for debugging to keep them in lowered form ?

File: base/compiler/optimize.jl
Comment (aviatesk) [reply base/compiler/optimize.jl]: What do you think about fea41f2654ac97d1ef59a7ac055c66e76b1bad57 ? We still need fresh optimization w/o inlining hack because those uncached sources are not optimized.

Well, now I'm wondering if it's really useful. I benchmarked some functions that hit this pass, and I couldn't find any cases where this "single-level-inlining" can help the performance.. Honestly I don't think it's worth the complexity that comes with this support.

**Comment by aviatesk** [comment]
Alright, I decided to give up the logic to keep otherwise-uncached source during abstractinterpret and re-optimize it to inline for now. To me it seems to be more trouble and complex than it's worth.
Still, with this PR we process statement flags before inference, and now  `const_prop_methodinstance_heuristic` respects statement-level `@inline` annotations, and so we can expect the annotations to work for most cases unless it's involved with recursive calls (I think the right solutions for them would be TCE and similar optimizations, rather than inlining).

So this PR ended up being quite simple at the end of the day, and I think it's ready to go.

File: base/compiler/optimize.jl
Comment (aviatesk) [reply base/compiler/optimize.jl]: https://github.com/JuliaLang/julia/pull/41328#issuecomment-904460017

**Review by JeffBezanson — APPROVED** [review]

File: base/compiler/optimize.jl
Comment (aviatesk) [reply base/compiler/optimize.jl:44]: It's possible, but I'm afraid to spam with such warnings (they're really not interest from user's point of view, I think):
```julia
module Pkg
recursivef(...) = ...
function userfacingf(...)
  @inline recursivef(...)
end
export userfacingf
end

userfacingf(...) # we want warning ... ?
```

We can have it as opt-in feature, but that could be addressed in another PR.
Or, we can create new JET analyzer to report such cases.

**Comment by aviatesk** [comment]
Rebased and added news entry for this.

Let me merge this tomorrow if there is no objection.

File: base/compiler/optimize.jl
Comment (aviatesk) [reply base/compiler/optimize.jl:44]: Okay, let's leave this to another PR.
Currently we don't have any agreement on how we want the compiler to emit warnings. E.g. it doesn't report anything back to us even if it finds obvious type errors during inference, but we could argue that it's more appropriate to get some warnings.

So I don't want this discussion to hold this PR.
If we adopt `@force_inline` naming, it should be done here before releasing it to nightly, but I'm afraid to accept it.
Given that there is no way to "force" inlining" with the existence of recursive calls, the essential difference between `@force_inline` and `@inline` is whether they emit the warning or not. And having two different names that does the same thing but the warning is more confusing than benefit imho.

**Comment by aviatesk** [comment]
It turns out this PR is very functional in actual execution, because inferred source is often discarded during codegen pipeline.
See https://github.com/JuliaLang/julia/issues/42078, and I may revert this PR for now.

**Comment by KristofferC** [comment]
is -> isn't?

**Comment by aviatesk** [comment]
ah, yeah thanks. silly typo...

**Comment by aviatesk** [comment]
https://github.com/JuliaLang/julia/pull/42082 should fix the "serious" flaw with this PR, and now callsite inlining should work more reliably.

**Comment by chriselrod** [comment]
@oscardssmith If you haven't tried this out yet:
```julia
julia> function vexp!(y,x)
           @inbounds for i ∈ eachindex(y,x)
               y[i] = exp(x[i])
           end
       end
vexp! (generic function with 1 method)

julia> function vexp_inline!(y,x)
           @inbounds for i ∈ eachindex(y,x)
               y[i] = @inline exp(x[i])
           end
       end
vexp_inline! (generic function with 1 method)

julia> x = randn(256); y = similar(x);

julia> @benchmark vexp!($y, $x)
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.234 μs …  2.336 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.329 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.334 μs ± 49.344 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                       █▄                                    ▁
  ▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅██▁▇▃▁▆▁▁▁▁▁▁▁▁▁▁▁▃▁▁▁▁▃▁▁▁▁▁▆▃▄▄▇▆▅▆ █
  1.23 μs      Histogram: log(frequency) by time     1.49 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark vexp_inline!($y, $x)
BenchmarkTools.Trial: 10000 samples with 540 evaluations.
 Range (min … max):  212.893 ns … 316.361 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     213.315 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   214.200 ns ±   6.480 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

   ▂▇█▇▃                   ▁▂▃▂▂▂▂                              ▂
  ▅█████▃▁▃▃▁▁▁▁▁▁▄▁▁▁▁▁▃▄▆████████▇██▇▇▇▆▅▄▃▅▄▅▃▁▄▄▃▄▄▄▄▄▅▃▄▄▅ █
  213 ns        Histogram: log(frequency) by time        220 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

**Pull request merged**
